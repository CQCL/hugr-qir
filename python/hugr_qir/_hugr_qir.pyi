def cli(*args: str) -> None: ...
def opt_level_choices() -> list[str]: ...
def compile_target_choices() -> list[str]: ...
def output_format_choices() -> list[str]: ...
