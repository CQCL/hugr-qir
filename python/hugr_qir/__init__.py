def main_cli():
    from ._hugr_qir import cli
    import sys
    cli(*sys.argv)
