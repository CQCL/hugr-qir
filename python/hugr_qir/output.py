from base64 import b64encode
from enum import Enum
from pathlib import Path

from llvmlite.binding import (  # type: ignore[import-untyped]
    create_context,
    parse_assembly,
)


class OutputFormat(Enum):
    LLVMIR = "llvmir"
    BITCODE = "bitcode"
    BASE64 = "base64"


def get_write_mode(out_format: OutputFormat | None, file_path: Path | None) -> str:
    if out_format and out_format in [OutputFormat.BITCODE, OutputFormat.BASE64]:
        return "wb"
    if file_path and file_path.suffix not in [".ll", ".asm"]:
        return "wb"
    return "w"


def ir_string_to_output_format(qir_ir: str, output_format: OutputFormat) -> str | bytes:
    match output_format:
        case OutputFormat.LLVMIR:
            return qir_ir
        case OutputFormat.BITCODE:
            ctx = create_context()
            module = parse_assembly(qir_ir, context=ctx)
            return module.as_bitcode()
        case OutputFormat.BASE64:
            ctx = create_context()
            module = parse_assembly(qir_ir, context=ctx)
            qir_bitcode = module.as_bitcode()
            return b64encode(qir_bitcode).decode("utf-8")
    errmsg = "Unrecognized output format"
    raise ValueError(errmsg)
