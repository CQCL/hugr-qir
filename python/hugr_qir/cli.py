"""Cli for hugr-qir."""

import tempfile
from importlib.metadata import version
from pathlib import Path

import click
from quantinuum_qircheck import qircheck

from hugr_qir._hugr_qir import cli


@click.command(name="hugr-qir")
@click.argument("hugr_file", type=click.Path(exists=True, path_type=Path))
@click.option(
    "--validate-qir/--no-validate-qir",
    "validate",
    default=True,
    help="Whether to validate output",
)
@click.option(
    "-o",
    "--output",
    "outfile",
    type=click.File("w"),
    default="-",
    help="Name of output file (optional)",
)
@click.version_option(version=version("hugr_qir"))
def hugr_qir(validate: bool, hugr_file: Path, outfile: click.File) -> None:
    """This is the cli for converting hugr to qir."""

    options = ["-q"]
    with tempfile.NamedTemporaryFile(delete=True, suffix=".ll") as temp_file:
        tmp_options = [*options, "-o", temp_file.name]
        cli(str(hugr_file), *tmp_options)
        with Path.open(Path(temp_file.name)) as output:
            qir = output.read()
    if validate:
        qircheck(qir)

    if hasattr(outfile, "write"):
        outfile.write(qir)
    else:
        # this should never happen but the check for write is
        # needed to appease mypy (click doesn't handle typing well)
        err_msg = f"Could not write to object {outfile}"
        raise AttributeError(err_msg)


if __name__ == "__main__":
    hugr_qir()
