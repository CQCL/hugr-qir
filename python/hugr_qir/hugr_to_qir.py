import tempfile
from pathlib import Path

from .cli import hugr_qir_impl


def hugr_to_qir(hugr: bytes, validate_qir: bool = True) -> str:
    """A function for converting hugr to qir."""
    with (
        tempfile.NamedTemporaryFile(delete=True, suffix=".hugr") as temp_infile,
        tempfile.NamedTemporaryFile(delete=True, suffix=".ll") as temp_outfile,
    ):
        with Path.open(Path(temp_infile.name), "wb") as cli_input:
            cli_input.write(hugr)
        with Path.open(Path(temp_outfile.name), "w") as cli_output:
            hugr_qir_impl(validate_qir, Path(cli_input.name), cli_output)
        with Path.open(Path(temp_outfile.name), "r") as cli_output:
            return cli_output.read()
