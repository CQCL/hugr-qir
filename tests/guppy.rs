use std::{env, path::Path, process::{self, Command, Stdio}};

use hugr::{package::Package, std_extensions::STD_REG, Hugr};
use hugr_llvm::inkwell;
use hugr_qir::CompileArgs;
use insta::assert_snapshot;
use itertools::Itertools as _;
use rstest::rstest;

fn capture_guppy(path: impl AsRef<Path>) -> (Hugr, String) {
    let process::Output {
        status,
        stdout,
        stderr,
    } = Command::new(".devenv/state/venv/bin/python")
        .arg(path.as_ref())
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();
    eprintln!("{}", std::str::from_utf8(&stdout).unwrap());
    eprintln!("{}", std::str::from_utf8(&stderr).unwrap());
    assert!(status.success());
    let hugr = Package::from_json_reader(stdout.as_slice(), &STD_REG)
        .unwrap()
        .modules
        .into_iter()
        .exactly_one()
        .unwrap();
    let stderr = String::from_utf8(stderr).unwrap();
    (hugr, stderr)
}

fn compile(hugr: &mut Hugr) -> String {
    let args = {
        let mut a = CompileArgs::default();
        a.validate = true;
        a
    };
    let context = inkwell::context::Context::create();
    let module = args.compile(hugr, &context).unwrap();
    module.to_string()
}

// This generates a test case for each file matching 'guppy_examples/**/*.py'
//
// For now we snapshot test the stderr output of executing the python file,
// and the LLVM IR output by hugr-qir.
//
// If we get problems with new test cases not being run in dev environments we might consider
// adding a `build.rs` as in:
// <https://docs.rs/rstest/latest/rstest/attr.rstest.html#files-path-as-input-arguments>
#[rstest]
fn guppy_examples(
    #[base_dir = "guppy_examples"]
    #[files("**/*.py")]
    file: impl AsRef<Path>,
) {
    let file2 = file.as_ref();
    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file2.file_stem().unwrap().to_str().unwrap());
    settings.bind(|| {
        let (mut hugr, stderr) = capture_guppy(file);
        assert_snapshot!("llvmir", compile(&mut hugr));
        assert_snapshot!("stderr", stderr);
    });
}
