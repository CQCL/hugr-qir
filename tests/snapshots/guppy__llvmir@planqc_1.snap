---
source: tests/guppy.rs
expression: compile(&mut hugr)
---
; ModuleID = 'hugr-qir'
source_filename = "hugr-qir"

%QUBIT = type opaque
%RESULT = type opaque

@0 = private unnamed_addr constant [47 x i8] c"tket2.rotation.from_halfturns_unchecked failed\00", align 1
@prelude.panic_template = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@1 = private unnamed_addr constant [37 x i8] c"Expected variant 1 but got variant 0\00", align 1
@prelude.panic_template.1 = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@2 = private unnamed_addr constant [2 x i8] c"1\00", align 1

define %QUBIT* @__hugr__.rx.1(%QUBIT* %0, { { double } } %1) {
alloca_block:
  %"0" = alloca %QUBIT*, align 8
  %"2_0" = alloca %QUBIT*, align 8
  %"2_1" = alloca { { double } }, align 8
  %"4_0" = alloca %QUBIT*, align 8
  %"6_0" = alloca %QUBIT*, align 8
  %"6_1" = alloca { { double } }, align 8
  %"03" = alloca %QUBIT*, align 8
  %"19_0" = alloca { {} }, align 8
  %"63_0" = alloca %QUBIT*, align 8
  %"61_0" = alloca double, align 8
  %"59_0" = alloca double, align 8
  %"57_0" = alloca double, align 8
  %"52_0" = alloca double, align 8
  %"46_0" = alloca double, align 8
  %"44_0" = alloca double, align 8
  %"47_0" = alloca %QUBIT*, align 8
  %"42_0" = alloca double, align 8
  %"48_0" = alloca %QUBIT*, align 8
  %"18_0" = alloca { {} }, align 8
  %"16_0" = alloca { {} }, align 8
  %"11_0" = alloca { {} }, align 8
  %"9_0" = alloca double, align 8
  %"12_0" = alloca { { double } }, align 8
  %"13_0" = alloca double, align 8
  %"14_0" = alloca double, align 8
  %"50_0" = alloca double, align 8
  %"53_0" = alloca double, align 8
  %"54_0" = alloca %QUBIT*, align 8
  %"62_0" = alloca %QUBIT*, align 8
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  store %QUBIT* %0, %QUBIT** %"2_0", align 8
  store { { double } } %1, { { double } }* %"2_1", align 8
  %"2_01" = load %QUBIT*, %QUBIT** %"2_0", align 8
  %"2_12" = load { { double } }, { { double } }* %"2_1", align 8
  store %QUBIT* %"2_01", %QUBIT** %"6_0", align 8
  store { { double } } %"2_12", { { double } }* %"6_1", align 8
  br label %2

2:                                                ; preds = %entry_block
  %"6_04" = load %QUBIT*, %QUBIT** %"6_0", align 8
  %"6_15" = load { { double } }, { { double } }* %"6_1", align 8
  store %QUBIT* %"6_04", %QUBIT** %"6_0", align 8
  store { { double } } %"6_15", { { double } }* %"6_1", align 8
  store double 0xBFF921FB54442D18, double* %"61_0", align 8
  store double 0x3FF921FB54442D18, double* %"59_0", align 8
  store double 0x400921FB54442D18, double* %"57_0", align 8
  store double 0x400921FB54442D18, double* %"52_0", align 8
  store double 0xBFF921FB54442D18, double* %"46_0", align 8
  store double 0x3FF921FB54442D18, double* %"44_0", align 8
  %"6_06" = load %QUBIT*, %QUBIT** %"6_0", align 8
  %"44_07" = load double, double* %"44_0", align 8
  %"46_08" = load double, double* %"46_0", align 8
  call void @__quantum__qis__phasedx__body(double %"44_07", double %"46_08", %QUBIT* %"6_06")
  store %QUBIT* %"6_06", %QUBIT** %"47_0", align 8
  store double 0x400921FB54442D18, double* %"42_0", align 8
  %"47_09" = load %QUBIT*, %QUBIT** %"47_0", align 8
  %"42_010" = load double, double* %"42_0", align 8
  call void @__quantum__qis__rz__body(double %"42_010", %QUBIT* %"47_09")
  store %QUBIT* %"47_09", %QUBIT** %"48_0", align 8
  store { {} } undef, { {} }* %"19_0", align 1
  store { {} } undef, { {} }* %"18_0", align 1
  store { {} } undef, { {} }* %"16_0", align 1
  store { {} } undef, { {} }* %"11_0", align 1
  %"6_111" = load { { double } }, { { double } }* %"6_1", align 8
  %3 = extractvalue { { double } } %"6_111", 0
  %4 = extractvalue { double } %3, 0
  store double %4, double* %"9_0", align 8
  %"9_012" = load double, double* %"9_0", align 8
  %5 = insertvalue { double } undef, double %"9_012", 0
  %6 = insertvalue { { double } } poison, { double } %5, 0
  store { { double } } %6, { { double } }* %"12_0", align 8
  %"12_013" = load { { double } }, { { double } }* %"12_0", align 8
  %7 = extractvalue { { double } } %"12_013", 0
  %8 = extractvalue { double } %7, 0
  store double %8, double* %"13_0", align 8
  %"13_014" = load double, double* %"13_0", align 8
  %9 = fcmp oeq double %"13_014", 0x7FF0000000000000
  %10 = fcmp oeq double %"13_014", 0xFFF0000000000000
  %11 = fcmp uno double %"13_014", 0.000000e+00
  %12 = or i1 %9, %10
  %13 = or i1 %12, %11
  %14 = xor i1 %13, true
  br i1 %14, label %18, label %16

15:                                               ; preds = %20
  %"029" = load %QUBIT*, %QUBIT** %"03", align 8
  store %QUBIT* %"029", %QUBIT** %"4_0", align 8
  %"4_030" = load %QUBIT*, %QUBIT** %"4_0", align 8
  store %QUBIT* %"4_030", %QUBIT** %"0", align 8
  %"031" = load %QUBIT*, %QUBIT** %"0", align 8
  ret %QUBIT* %"031"

16:                                               ; preds = %2
  %17 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template, i32 0, i32 0), i32 1, i8* getelementptr inbounds ([47 x i8], [47 x i8]* @0, i32 0, i32 0))
  call void @abort()
  unreachable

18:                                               ; preds = %2
  store double %"13_014", double* %"14_0", align 8
  %"14_015" = load double, double* %"14_0", align 8
  store double %"14_015", double* %"50_0", align 8
  %"50_016" = load double, double* %"50_0", align 8
  %"52_017" = load double, double* %"52_0", align 8
  %19 = fmul double %"50_016", %"52_017"
  store double %19, double* %"53_0", align 8
  %"48_018" = load %QUBIT*, %QUBIT** %"48_0", align 8
  %"53_019" = load double, double* %"53_0", align 8
  call void @__quantum__qis__rz__body(double %"53_019", %QUBIT* %"48_018")
  store %QUBIT* %"48_018", %QUBIT** %"54_0", align 8
  %"54_020" = load %QUBIT*, %QUBIT** %"54_0", align 8
  %"59_021" = load double, double* %"59_0", align 8
  %"61_022" = load double, double* %"61_0", align 8
  call void @__quantum__qis__phasedx__body(double %"59_021", double %"61_022", %QUBIT* %"54_020")
  store %QUBIT* %"54_020", %QUBIT** %"62_0", align 8
  %"62_023" = load %QUBIT*, %QUBIT** %"62_0", align 8
  %"57_024" = load double, double* %"57_0", align 8
  call void @__quantum__qis__rz__body(double %"57_024", %QUBIT* %"62_023")
  store %QUBIT* %"62_023", %QUBIT** %"63_0", align 8
  %"19_025" = load { {} }, { {} }* %"19_0", align 1
  %"63_026" = load %QUBIT*, %QUBIT** %"63_0", align 8
  store { {} } %"19_025", { {} }* %"19_0", align 1
  store %QUBIT* %"63_026", %QUBIT** %"63_0", align 8
  %"19_027" = load { {} }, { {} }* %"19_0", align 1
  %"63_028" = load %QUBIT*, %QUBIT** %"63_0", align 8
  switch i32 0, label %20 [
  ]

20:                                               ; preds = %18
  %21 = extractvalue { {} } %"19_027", 0
  store %QUBIT* %"63_028", %QUBIT** %"03", align 8
  br label %15
}

declare void @__quantum__qis__phasedx__body(double, double, %QUBIT*)

declare void @__quantum__qis__rz__body(double, %QUBIT*)

declare i32 @printf(i8*, ...)

declare void @abort()

define void @__hugr__.main.20() {
alloca_block:
  %"37_0" = alloca { {} }, align 8
  %"64_0" = alloca { i32, {}, { %QUBIT* } }, align 8
  %"65_0" = alloca %QUBIT*, align 8
  %"0" = alloca %QUBIT*, align 8
  %"70_0" = alloca { i32, i8* }, align 8
  %"71_0" = alloca %QUBIT*, align 8
  %"05" = alloca %QUBIT*, align 8
  %"73_0" = alloca %QUBIT*, align 8
  %"36_0" = alloca { {} }, align 8
  %"33_0" = alloca { {} }, align 8
  %"30_0" = alloca double, align 8
  %"31_0" = alloca { { double } }, align 8
  %"32_0" = alloca %QUBIT*, align 8
  %"34_0" = alloca { i32, {}, {} }, align 8
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  br label %0

0:                                                ; preds = %entry_block
  %1 = call %QUBIT* @__quantum__rt__qubit_allocate()
  %2 = insertvalue { %QUBIT* } undef, %QUBIT* %1, 0
  %3 = insertvalue { i32, {}, { %QUBIT* } } { i32 1, {} poison, { %QUBIT* } poison }, { %QUBIT* } %2, 2
  store { i32, {}, { %QUBIT* } } %3, { i32, {}, { %QUBIT* } }* %"64_0", align 8
  %"64_01" = load { i32, {}, { %QUBIT* } }, { i32, {}, { %QUBIT* } }* %"64_0", align 8
  %4 = extractvalue { i32, {}, { %QUBIT* } } %"64_01", 0
  switch i32 %4, label %5 [
    i32 1, label %7
  ]

5:                                                ; preds = %0
  %6 = extractvalue { i32, {}, { %QUBIT* } } %"64_01", 1
  br label %cond_65_case_0

7:                                                ; preds = %0
  %8 = extractvalue { i32, {}, { %QUBIT* } } %"64_01", 2
  %9 = extractvalue { %QUBIT* } %8, 0
  store %QUBIT* %9, %QUBIT** %"05", align 8
  br label %cond_65_case_1

10:                                               ; preds = %22
  ret void

cond_65_case_0:                                   ; preds = %5
  store { i32, i8* } { i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @1, i32 0, i32 0) }, { i32, i8* }* %"70_0", align 8
  %"70_03" = load { i32, i8* }, { i32, i8* }* %"70_0", align 8
  %11 = extractvalue { i32, i8* } %"70_03", 0
  %12 = extractvalue { i32, i8* } %"70_03", 1
  %13 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template.1, i32 0, i32 0), i32 %11, i8* %12)
  call void @abort()
  store %QUBIT* null, %QUBIT** %"71_0", align 8
  %"71_04" = load %QUBIT*, %QUBIT** %"71_0", align 8
  store %QUBIT* %"71_04", %QUBIT** %"0", align 8
  br label %cond_exit_65

cond_65_case_1:                                   ; preds = %7
  %"06" = load %QUBIT*, %QUBIT** %"05", align 8
  store %QUBIT* %"06", %QUBIT** %"73_0", align 8
  %"73_07" = load %QUBIT*, %QUBIT** %"73_0", align 8
  store %QUBIT* %"73_07", %QUBIT** %"0", align 8
  br label %cond_exit_65

cond_exit_65:                                     ; preds = %cond_65_case_1, %cond_65_case_0
  %"02" = load %QUBIT*, %QUBIT** %"0", align 8
  store %QUBIT* %"02", %QUBIT** %"65_0", align 8
  store { {} } undef, { {} }* %"37_0", align 1
  store { {} } undef, { {} }* %"36_0", align 1
  store { {} } undef, { {} }* %"33_0", align 1
  store double 1.500000e+00, double* %"30_0", align 8
  %"30_08" = load double, double* %"30_0", align 8
  %14 = insertvalue { double } undef, double %"30_08", 0
  %15 = insertvalue { { double } } poison, { double } %14, 0
  store { { double } } %15, { { double } }* %"31_0", align 8
  %"65_09" = load %QUBIT*, %QUBIT** %"65_0", align 8
  %"31_010" = load { { double } }, { { double } }* %"31_0", align 8
  %16 = call %QUBIT* @__hugr__.rx.1(%QUBIT* %"65_09", { { double } } %"31_010")
  store %QUBIT* %16, %QUBIT** %"32_0", align 8
  %"32_011" = load %QUBIT*, %QUBIT** %"32_0", align 8
  %17 = call %RESULT* @__quantum__qis__m__body(%QUBIT* %"32_011")
  %18 = call i1 @__quantum__qis__read_result__body(%RESULT* %17)
  %19 = select i1 %18, { i32, {}, {} } { i32 1, {} poison, {} undef }, { i32, {}, {} } { i32 0, {} undef, {} poison }
  store { i32, {}, {} } %19, { i32, {}, {} }* %"34_0", align 4
  %"34_012" = load { i32, {}, {} }, { i32, {}, {} }* %"34_0", align 4
  %20 = extractvalue { i32, {}, {} } %"34_012", 0
  %21 = trunc i32 %20 to i1
  call void @__quantum__rt__bool_record_output(i1 %21, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @2, i32 0, i32 0))
  %"37_013" = load { {} }, { {} }* %"37_0", align 1
  store { {} } %"37_013", { {} }* %"37_0", align 1
  %"37_014" = load { {} }, { {} }* %"37_0", align 1
  switch i32 0, label %22 [
  ]

22:                                               ; preds = %cond_exit_65
  %23 = extractvalue { {} } %"37_014", 0
  br label %10
}

declare %QUBIT* @__quantum__rt__qubit_allocate()

declare %RESULT* @__quantum__qis__m__body(%QUBIT*)

declare i1 @__quantum__qis__read_result__body(%RESULT*)

declare void @__quantum__rt__bool_record_output(i1, i8*)
